### PR TITLE
Classification

### PR DESCRIPTION
**Pull request by diegomarvid**
- [x] Add support for classification pipelines

### CODE REVIEW

File: ml_garden/core/pipeline.py [hunk 1]
```diff
@@ -32,10 +32,13 @@ class Pipeline:
         "explainer",
     ]
 
+    SUPPORTED_TASKS = {"classification", "regression"}
```
Comment (sebastian-correa) [inline ml_garden/core/pipeline.py:35]: I think it's a good moment to make this a `str, Enum`. The key advantage is that it allows good type-hints (as `task: SupportedTasks` is more descriptive than `task: str`). If this is done, I'd declare the enum quite top-level and import it everywhere where it's needed.

At the very least, I'd use `Literal` in the type hints instead of bare `str`, though that necessitates more maintenance.

File: ml_garden/core/steps/calculate_metrics.py [hunk 2]
```diff
@@ -19,8 +29,8 @@ def __init__(self) -> None:
         super().__init__()
         self.init_logger()
 
-    def _calculate_metrics(self, true_values: pd.Series, predictions: pd.Series) -> dict:
-        """Calculate metrics.
+    def _calculate_regression_metrics(self, true_values: pd.Series, predictions: pd.Series) -> dict:
```
Comment (sebastian-correa) [inline ml_garden/core/steps/calculate_metrics.py:32]: It _might_ be a good idea to have some `_calculate_metrics` that takes `task: SupportedTasks` and calls the appropriate method. This allows users to be more programmatic with these calls, though I'm not sure it's a usecase ML Garden users would face at all.

File: ml_garden/core/steps/calculate_metrics.py [hunk 3]
```diff
@@ -50,6 +60,56 @@ def _calculate_metrics(self, true_values: pd.Series, predictions: pd.Series) ->
...
+        return {
+            "Overall": {
+                "Accuracy": str(accuracy),
+                "Weighted Precision": str(weighted_precision),
+                "Weighted Recall": str(weighted_recall),
+                "Weighted F1 Score": str(weighted_f1),
+            },
+            "Per_Class": class_metrics,
+            "Confusion Matrix": str(cm.tolist()),
+        }
```
Comment (sebastian-correa) [inline ml_garden/core/steps/calculate_metrics.py:111]: We _could_ use `TypedDict`s or `dataclasses` here so that the structure of the metrics is self-documenting.

File: ml_garden/core/steps/fit_model.py [hunk 4]
```diff
@@ -124,7 +131,7 @@ def _create_study(self) -> optuna.Study:
         return study
 
     @staticmethod
-    def _calculate_error(y_true, y_pred, metric):
+    def _calculate_error(y_true: np.ndarray, y_pred: np.ndarray, metric: str, task: str) -> float:
```
Comment (sebastian-correa) [inline ml_garden/core/steps/fit_model.py:134]: Not sure about the library, but maybe you could support `numpy.typing.ArrayLike`s instead of just `np.ndarray`s, since most numpy functions actually support being called with other things (like `list`s).

This is very low priority I'd say, but it could help if passing a `pd.Series`.

File: ml_garden/core/model.py [hunk 5]
```diff
@@ -3,11 +3,13 @@
...
 import pandas as pd
 
+from ml_garden.core.constants import Task
+
 
 class Model(ABC):
     """Base class for models."""
 
-    TASK = ""
+    TASKS: List[Task] = []
```
Comment (sebastian-correa) [inline ml_garden/core/model.py:12]: This was a simple `str` before. Are you sure you need a `List[Task]` or just `Task`?

File: ml_garden/core/pipeline.py [hunk 6]
```diff
@@ -146,6 +144,14 @@ def _validate_configuration(cls, config: dict[str, Any]) -> None:
                 ' "parameters" section in the config.'
             )
 
+    @classmethod
+    def _validate_task(cls, task: str) -> Task:
+        try:
+            return Task(task.lower())
```
Comment (sebastian-correa) [inline ml_garden/core/pipeline.py:150]: This is preference, but we usually favor enum _names_ over _values_ (as values are just arbitrary, see the fact that `Enum` has `auto()` for auto populating values).

This translates to looking up values from config files with `[]` instead of `()`.

However, if you maintain the usual `SOMETHING = "something"` pattern of `str, Enum`s and use `.lower()` to compare, then this comment is completely moot.

The only relevant thing is that users know what to write in their settings. Is it the values of enum members? Is it the names? Does it not matter?

File: ml_garden/core/constants.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
+from enum import Enum
+
+
+class Task(Enum):
```
Comment (sebastian-correa) [inline ml_garden/core/constants.py:4]: I think this should be turned to a `str, Enum`. It makes things more interoperable and has no effect in your current implementation :). This is because enum _members_ become `str`s as well (i.e., `isinstance(Task.CLASSIFICATION, str) == True`) where they behave as their values. You can treat them as strings in many places without too much issue.

For example, users could easily pass `"classification"` and it should work, so you wouldn't need to do much conversion from setting `str` to `Enum` member. If you want to do it to validate the value, you could do this:

```python
task = Task(task)
```

1. If you get `task` as an enum member, it won't fail and it'll return itself. Remember, the enum member is a str, so `()` lookup (by value) works fine.
2. If you get `task` as a `str`, it'll fail only if the `str` is not a valid task (the `()` lookup fails). Otherwise, you get the enum member.

So you get free validation and some more flexibility.

You ca test this out by doing
```python
class A(str, Enum):
    A = "a"
     B = "b"

print(A("a"))
print(A(A.A))
print(A("other"))
```

**Review by sebastian-correa — COMMENTED** [review]
Some comments on the `Task` `Enum`.

File: ml_garden/core/model.py (hunk 5 above)
Comment (diegomarvid) [reply ml_garden/core/model.py:12]: For most cases I need just a Task. For example: XGBRegressor or XGBClassifier.

But I was adding support for [AutoGluon](https://auto.gluon.ai/stable/index.html) on another PR and that is an auto model that works with more than one Task. I don't know if there are more models that support more than one task but it was my way for fixing this for AutoGluon

File: ml_garden/implementation/tabular/xgboost/model.py [hunk 8]
```diff
@@ -1,79 +1,17 @@
...
+from ml_garden.core.constants import Task
 from ml_garden.core.model import Model
-from ml_garden.utils import compression_utils
 
 logger = logging.getLogger(__file__)
 
 
-# %%
-# Function to save and compress an XGBoost model
-def save_and_compress_xgb_booster(
```
[the remark this answers is not in the document]
Comment (Ludecan) [reply ml_garden/implementation/tabular/xgboost/model.py:18]: Maybe we could have an `XGBoostBase` class. It would have the fit and predict methods which are the same between XGBoostClassifier and XGBoostRegressor and move these two as methods of the class.

**Review by Ludecan — COMMENTED** [review]
@diegomarvid , this is awesome!
So cool to see classification support in the lib.
I left a few comments for you to check out.
Maybe it's worth having some classification example configs thrown in too to exemplify how it works?

File: ml_garden/core/steps/explainer_dashboard.py [hunk 9]
```diff
@@ -97,12 +96,22 @@ def execute(self, data: DataContainer) -> DataContainer:
             # To avoid this potential long wait to a crash we add this assertion here
             assert len(X) == len(y), "Mismatch in number of samples and labels"
 
-            explainer = RegressionExplainer(
+            # Choose the appropriate explainer based on the task
+            explainer_class = {
+                Task.REGRESSION: RegressionExplainer,
+                Task.CLASSIFICATION: ClassifierExplainer,
+            }.get(data.task)
```
Comment (Ludecan) [inline ml_garden/core/steps/explainer_dashboard.py:103]: Beautiful!

File: ml_garden/core/steps/fit_model.py [hunk 10]
```diff
@@ -2,14 +2,19 @@
...
 from typing import Optional, Type
 
+import numpy as np
 import optuna
 from sklearn.metrics import (
+    accuracy_score,
+    f1_score,
     mean_absolute_error,
     mean_squared_error,
+    roc_auc_score,
```
[the remark this answers is not in the document]
Comment (Ludecan) [reply ml_garden/core/steps/fit_model.py:12]: Created the issue [here](https://github.com/tryolabs/ml-garden/issues/34)

File: ml_garden/core/steps/encode.py [hunk 11]
```diff
@@ -277,70 +277,76 @@ def _get_encoder_class_and_params(
...
-        if not self.feature_encoders:
-            high_cardinality_encoder_class, high_cardinality_encoder_params = (
-                self._get_encoder_class_and_params(self.HIGH_CARDINALITY_ENCODER)
-            )
-            low_cardinality_encoder_class, low_cardinality_encoder_params = (
-                self._get_encoder_class_and_params(self.LOW_CARDINALITY_ENCODER)
+        encoded_features = set()
+
+        # Handle shared encoder for Origin and Destination
+        if "shared_origin_destination" in self.feature_encoders:
```
Comment (Ludecan) [inline ml_garden/core/steps/encode.py:308]: Hmmm, this feels a bit too specific. Like origin and destination are the columns in this particular dataset right?
Do we even need to share encoding between the columns?
Either if we use OrdinalEncoding or TargetEncoding, the tree will learn the cuts it needs to do for each encode separately and there would be no difference right? Only the cut thresholds but the results should be the same?

File: tests/core/test_pipeline.py [hunk 12]
```diff
@@ -17,6 +18,15 @@ def test_simple_train_pipeline():
     os.remove(zip_file)
 
 
+def test_simple_train_pipeline_classification():
+    """Test that the pipeline can be trained with classification."""
+    pipeline = Pipeline.from_json("tests/data/test_classification.json")
+    pipeline.train()
```
Comment (Ludecan) [inline tests/core/test_pipeline.py:24]: 💯

**Review by Ludecan — APPROVED** [review]
Looking good @diegomarvid !! Solid work here!
